### DISCUSSION TITLE
Add marks for clearing and other semantic functionality

### DISCUSSION THREAD

**Discussion by edwardloveall** [ghostty-org/ghostty#3469] [Feature Requests, Ideas]
One thing I miss from Apple's Terminal.app is the ability to clear to the previous prompt. They enable with with the concept of [marks](https://blog.edwardloveall.com/terminal-marks). I use this all the time when I've just printed a mistake to the screen and want to remove just that. A mistyped command, a large amount of output that I go on to refine, etc. It could also be used to copy the output of the last command, jump backwards and forwards by command (instead of some number of lines), and create arbitrary points in a large printed log.

## Related

* [OSC 133: Support semantic prompt regions](https://github.com/ghostty-org/ghostty/issues/5932)
* [Select output of last command](https://github.com/ghostty-org/ghostty/discussions/4385)
* [Is there a way to clear screen and save contents in scrollback](https://github.com/ghostty-org/ghostty/discussions/5269)

**Comment by dkmar** [comment]
Sweet, I'll move my post here. Gauging interest in bringing the macOS Terminal's "Clear to Previous Mark" feature to Ghostty. If you're not familiar, it clears just the last command:

https://github.com/user-attachments/assets/e4757b5b-64bb-45b7-9fa2-667cf8166455

### Use Cases
1. Clearing mistyped commands from the screen.
2. Clearing verbose commands like `ghostty +list-keybinds`.  This use case is partially supplanted by having a quick terminal to run stuff without polluting my current terminal screen, but I still think it's useful here.

### Precedent
This feature is implemented by Terminal and [iTerm2](https://gitlab.com/gnachman/iterm2/-/issues/7343). 
We already have semantic prompt marking and a `jump_to_prompt` feature in Ghostty which is analogous to the "Jump to Previous/Next Mark" navigation commands in Terminal.

### Misc
[What is Cmd+L?](https://old.reddit.com/r/commandline/comments/a29fqc/how_to_get_cmd_l_clear_line_in_iterm/eax01g2/) 
[Marks in the macOS Terminal](https://apple.stackexchange.com/questions/209635/what-functionality-do-marks-offer-in-the-el-capitan-terminal)

**Comment by dwt** [comment]
I definitely want this too. I use that all the time to correct for mistakes and leave only the good history, or to try something to see if it's helpful.

**Comment by okuuva** [comment]
@edwardloveall would it be ok to change this FR to be about marks in general? I was about to create a new discussion on the topic but it seems to me that there's plenty of context here already.

In addition to enabling clearing the output of the previous prompt, marks would also enable a more general solution to selecting the output of previous command (see #4385) like @indirect [pointed out](https://github.com/ghostty-org/ghostty/discussions/4385#discussioncomment-12887287). They would also potentially enable stuff like clearing the screen while preserving scrollback (see #5269).

**Reply by edwardloveall** [reply]
Yeah, I'd be fine with that. I'll change the title and description.

**Reply by dkmar** [reply]
Always happy to have another bump on this. There’s also #5932 that might be useful for Clearing to Previous Prompt and Selecting Previous Output

**Reply by edwardloveall** [reply]
Thanks. I'll add some related items to the description.

**Comment by erickrf** [comment]
I would say this is the one feature that makes me stick to Terminal.app. Is there maybe a workaround to emulate this in Ghostty?

**Reply by indirect** [reply]
I ended up configuring a workaround inside tmux and using that instead. That said, I'm still interested in something directly in Ghostty that would let me select and/or copy the most recent command and its output.

**Comment by cmpute** [comment]
Since it's related, I'm posting my few cents here from https://github.com/ghostty-org/ghostty/discussions/9706. The integrated terminal of vscode also support some nice semantic features. For full ref see: https://code.visualstudio.com/docs/terminal/shell-integration

Briefly, vscode's terminal also group terminal text by commands (similar to the `mark` concept here, although I'm not sure whether they are exact the same thing). It provides two thing that's very handy (see the screenshot below): a decoration near every prompt, and a highlight in the scrollbar.

<img width="1952" height="850" alt="image" src="https://github.com/user-attachments/assets/b6b51718-78fe-4ea5-ac92-5a3153e051e6" />

The decorators (blue/red dots on the left) allows one to copy the exit code, copy the command output and even rerun the command by  clicking the dot.

The scrollbar will highlight the status of the command (blue/red markups on the right), which is also helpful for navigating between commands.

Furthermore, vscode has a ['sticky scroll'](https://code.visualstudio.com/docs/terminal/shell-integration#_sticky-scroll) feature, which pins the command input to the top of the panel. This is a not-important-but-nice-to-have feature to me.
